desub/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+// along with substrate-desub.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Facade crate for decoding data that uses any version of metadata (V8+)
+
+#![forbid(unsafe_code)]
+#[deny(unused)]
+
+mod error;
+
+use codec::Decode;
+use core_v14::{decoder::Extrinsic, metadata::runtime_metadata_version, Decoder as TypeInfoDecoder, Metadata as DesubMetadata};
+use desub_legacy::{
+	decoder::{Decoder as LegacyDecoder, Metadata as LegacyDesubMetadata},
+	RustTypeMarker, TypeDetective,
+};
+use serde_json::Value;
+use frame_metadata::RuntimeMetadataPrefixed;
+use std::{collections::HashMap, marker::PhantomData};
+
+#[cfg(feature = "polkadot-js")]
+use extras::{TypeResolver as PolkadotJsResolver};
+
+#[cfg(feature = "polkadot-js")]
+pub use extras::runtimes;
+pub use desub_legacy::decoder::Chain;
+pub use desub_common::SpecVersion;
+pub use self::error::Error;
+
+/// Struct That implements TypeDetective but refuses to resolve anything
+/// that is not of metadata v14+.
+/// Useful for use with a new chain that does not require historical metadata.
+#[derive(Copy, Clone, Debug)]
+struct NoLegacyTypes;
+
+impl TypeDetective for NoLegacyTypes {
+	fn get(&self, _: &str, _: u32, _: &str, _: &str) -> Option<&RustTypeMarker> {
+		None
+	}
+
+	fn try_fallback(&self, _: &str, _: &str) -> Option<&RustTypeMarker> {
+		None
+	}
+
+	fn get_extrinsic_ty(&self, _: &str, _: u32, _: &str) -> Option<&RustTypeMarker> {
+		None
+	}
+}
+
+pub struct Decoder<T: TypeDetective> {
+	legacy_decoder: LegacyDecoder,
+	current_decoder: HashMap<SpecVersion, TypeInfoDecoder>,
+	_marker: PhantomData<T>,
+}
+
+impl<T: TypeDetective> Decoder<T> {
+
+	/// Create a new general Decoder
+	pub fn new(types: impl TypeDetective + 'static, chain: Chain) -> Self {
+		let legacy_decoder = LegacyDecoder::new(types, chain);
+		let current_decoder = HashMap::new();
+		Self { legacy_decoder, current_decoder, _marker: PhantomData }
+	}
+
+	/// Register a runtime version with the decoder.
+	pub fn register_version(&mut self, version: SpecVersion, mut metadata: &[u8]) -> Result<(), Error> {
+		let metadata: RuntimeMetadataPrefixed = Decode::decode(&mut metadata)?;
+		if runtime_metadata_version(&metadata.1) >= 14 {
+			let meta = DesubMetadata::from_runtime_metadata(metadata.1)?;
+			let decoder = TypeInfoDecoder::with_metadata(meta);
+			self.current_decoder.insert(version, decoder);
+		} else {
+			self.legacy_decoder.register_version(version, LegacyDesubMetadata::from_runtime_metadata(metadata.1)?)?;
+		}
+		Ok(())
+	}
+
+	pub fn decode_extrinsics(&self, version: SpecVersion, mut data: &[u8]) -> Result<Value, Error> {
+		if self.current_decoder.contains_key(&version) {
+			let decoder = self.current_decoder.get(&version).expect("Checked if key is contained; qed");
+			match decoder.decode_extrinsics(&mut data) {
+				Ok(v) => Ok(serde_json::to_value(&v)?),
+				Err((ext, e)) => Err(Error::V14{ source: e, ext: ext.into_iter().map(Extrinsic::into_owned).collect() }),
```

Comment:
This is just because we might hit an error after some amount of extrinsics, and so I wanted to return all of the extrinsics parsed so far. If you like, you can just discard the extrinsics and report the error for the block, or we know which extrinsic failed by counting the successful ones?

I dont' think it makes sense to embed them into the error, because it wouldn't make sense for most errors.

I did try having the `decode_extrinsics` function return an iterator (which would allow you to collect extrinsics and stop when you hit an error (which would leave you with the same result), but it would have added a bit much complexity so I scrapped it!